File: cf_proxy.py
#!/usr/bin/env python3
"""
Cloudflare Access Cookie Proxy
Sits between client apps and Cloudflare Access protected services
Handles cookie authentication transparently
"""

import requests
import threading
from http.server import HTTPServer, BaseHTTPRequestHandler
import urllib.parse
import json
import time
from typing import Optional

class CloudflareAccessProxy:
    def __init__(self, target_domain: str, local_port: int = 8081):
        self.target_domain = target_domain
        self.local_port = local_port
        self.session = requests.Session()
        self.cf_cookie: Optional[str] = None
        self.cookie_expires = 0

    def extract_cf_cookie_from_browser(self) -> bool:
        """
        Attempts to extract CF_Authorization cookie from browser
        You'd need to implement browser cookie extraction here
        For now, returns False to trigger manual auth
        """
        # TODO: Implement browser cookie extraction
        # Could use browser_cookie3 library or similar
        return False

    def authenticate_manual(self) -> bool:
        """
        Manual authentication - user provides cookie
        """
        print(f"\n🔐 Authentication required for {self.target_domain}")
        print("1. Open your browser and go to:")
        print(f"   https://{self.target_domain}")
        print("2. Complete Cloudflare authentication")
        print("3. Open browser dev tools → Application → Cookies")
        print(f"4. Find 'CF_Authorization' cookie for {self.target_domain}")
        print("5. Copy the cookie value and paste it here:")

        cookie_value = input("\nCF_Authorization cookie: ").strip()

        if cookie_value:
            self.cf_cookie = cookie_value
            # Set expiration to 23 hours from now (typical CF Access session)
            self.cookie_expires = time.time() + (23 * 3600)
            return True
        return False

    def is_authenticated(self) -> bool:
        """Check if we have a valid authentication cookie"""
        return (self.cf_cookie is not None and
                time.time() < self.cookie_expires)

    def ensure_authenticated(self) -> bool:
        """Ensure we have valid authentication"""
        if self.is_authenticated():
            return True

        # Try to get cookie from browser first
        if self.extract_cf_cookie_from_browser():
            return True

        # Fall back to manual authentication
        return self.authenticate_manual()

class ProxyHandler(BaseHTTPRequestHandler):
    def __init__(self, proxy_instance, *args, **kwargs):
        self.proxy = proxy_instance
        super().__init__(*args, **kwargs)

    def do_GET(self):
        self._proxy_request('GET')

    def do_POST(self):
        self._proxy_request('POST')

    def do_PUT(self):
        self._proxy_request('PUT')

    def do_DELETE(self):
        self._proxy_request('DELETE')

    def do_HEAD(self):
        self._proxy_request('HEAD')

    def _proxy_request(self, method: str):
        """Proxy the request to the target domain with CF authentication"""

        # Ensure we're authenticated
        if not self.proxy.ensure_authenticated():
            self.send_error(401, "Authentication failed")
            return

        # Build target URL
        target_url = f"https://{self.proxy.target_domain}{self.path}"

        # Prepare headers
        headers = {}
        for header, value in self.headers.items():
            # Skip host header as we're changing the target
            # Also skip accept-encoding to avoid compression issues
            if header.lower() not in ['host', 'accept-encoding']:
                headers[header] = value

        # Fix OCI manifest support for registry requests
        if self.path.startswith('/v2/') and ('Accept' in headers):
            # Always add OCI manifest support for registry API calls
            oci_types = [
                'application/vnd.oci.image.manifest.v1+json',
                'application/vnd.oci.image.index.v1+json'
            ]
            current_accept = headers['Accept']
            for oci_type in oci_types:
                if oci_type not in current_accept:
                    headers['Accept'] += f', {oci_type}'

        # Add Cloudflare Access cookie
        if 'Cookie' in headers:
            headers['Cookie'] += f"; CF_Authorization={self.proxy.cf_cookie}"
        else:
            headers['Cookie'] = f"CF_Authorization={self.proxy.cf_cookie}"

        try:
            # Read request body if present
            content_length = int(self.headers.get('Content-Length', 0))
            body = self.rfile.read(content_length) if content_length > 0 else None

            # Make the proxied request
            response = self.proxy.session.request(
                method=method,
                url=target_url,
                headers=headers,
                data=body,
                stream=True,
                allow_redirects=False
            )

            # Send response status
            self.send_response(response.status_code)

            # Forward response headers
            for header, value in response.headers.items():
                # Skip headers that could cause issues
                if header.lower() not in ['transfer-encoding', 'connection']:
                    self.send_header(header, value)
            self.end_headers()

            # Forward response body
            if method != 'HEAD':
                for chunk in response.iter_content(chunk_size=8192):
                    if chunk:
                        self.wfile.write(chunk)

        except Exception as e:
            print(f"❌ Proxy error: {e}")
            self.send_error(500, f"Proxy error: {str(e)}")

    def log_message(self, format, *args):
        """Custom logging"""
        print(f"🔄 {format % args}")

def create_handler(proxy_instance):
    """Create handler class with proxy instance"""
    def handler(*args, **kwargs):
        return ProxyHandler(proxy_instance, *args, **kwargs)
    return handler

def main():
    import argparse

    parser = argparse.ArgumentParser(description='Cloudflare Access Cookie Proxy')
    parser.add_argument('domain', help='Target domain (e.g., registry.yourdomain.com)')
    parser.add_argument('--port', type=int, default=8081, help='Local proxy port')

    args = parser.parse_args()

    print(f"🚀 Starting Cloudflare Access Proxy")
    print(f"📡 Target: {args.domain}")
    print(f"🔗 Local proxy: http://localhost:{args.port}")

    # Create proxy instance
    proxy = CloudflareAccessProxy(args.domain, args.port)

    # Authenticate upfront before starting server
    print("\n🔐 Authentication required before starting proxy...")
    if not proxy.ensure_authenticated():
        print("❌ Authentication failed. Exiting.")
        return

    print("✅ Authentication successful!")

    # Create HTTP server
    server = HTTPServer(('localhost', args.port), create_handler(proxy))

    print(f"\n🚀 Proxy running on http://localhost:{args.port}")
    print(f"💡 Configure your client to use localhost:{args.port} instead of {args.domain}")
    print("🛑 Press Ctrl+C to stop")

    try:
        server.serve_forever()
    except KeyboardInterrupt:
        print("\n👋 Shutting down proxy...")
        server.shutdown()

if __name__ == "__main__":
    main()

File: cf_proxy.go
package main

import (
    "bufio"
    "bytes"
    "fmt"
    "io"
    "net/http"
    "os"
    "os/exec"
    "strings"
    "time"
    "strconv"

    "github.com/sirupsen/logrus"
)

// CloudflareAccessProxy represents the proxy instance
type CloudflareAccessProxy struct {
    targetDomain   string
    localPort      int
    session        *http.Client
    cfCookie       string
    cookieExpires  int64
    logger         *logrus.Logger
}

// NewCloudflareAccessProxy creates a new proxy instance
func NewCloudflareAccessProxy(targetDomain string, localPort int) *CloudflareAccessProxy {
    return &CloudflareAccessProxy{
        targetDomain:  targetDomain,
        localPort:     localPort,
        session:       &http.Client{Timeout: 30 * time.Second},
        logger:        logrus.New(),
    }
}

// authenticateWithCloudflared authenticates using cloudflared CLI if available
func (p *CloudflareAccessProxy) authenticateWithCloudflared() bool {
    p.logger.Info("🔐 Authenticating with cloudflared...")
    p.logger.Infof("📡 Target domain: %s", p.targetDomain)

    // Check if cloudflared is available
    if _, err := exec.LookPath("cloudflared"); err != nil {
        p.logger.Error("❌ cloudflared not found in PATH")
        return false
    }

    // Login with cloudflared
    p.logger.Info("🚀 Launching cloudflared access login...")
    loginCmd := exec.Command("cloudflared", "access", "login", fmt.Sprintf("https://%s", p.targetDomain))
    loginCmd.Stdout = os.Stdout
    loginCmd.Stderr = os.Stderr

    if err := loginCmd.Run(); err != nil {
        p.logger.Errorf("❌ cloudflared login failed: %v", err)
        return false
    }

    // Get the token
    p.logger.Info("🔑 Retrieving access token...")
    tokenCmd := exec.Command("cloudflared", "access", "token", fmt.Sprintf("-app=https://%s", p.targetDomain))
    tokenCmd.Stderr = os.Stderr

    output, err := tokenCmd.Output()
    if err != nil {
        p.logger.Errorf("❌ Failed to retrieve token: %v", err)
        return false
    }

    token := strings.TrimSpace(string(output))
    if token != "" {
        p.cfCookie = token
        // Set expiration to 23 hours from now (typical CF Access session)
        p.cookieExpires = time.Now().Unix() + int64(23*3600)
        p.logger.Info("✅ Successfully authenticated via cloudflared!")
        return true
    }

    return false
}

// extractCFCookieFromBrowser attempts to extract CF_Authorization cookie from browser
func (p *CloudflareAccessProxy) extractCFCookieFromBrowser() bool {
    // TODO: Implement browser cookie extraction
    // Could use a library like github.com/tebeka/selenium or similar
    p.logger.Info("⚠️  Browser cookie extraction not implemented")
    return false
}

// authenticateManual handles manual authentication
func (p *CloudflareAccessProxy) authenticateManual() bool {
    p.logger.Infof("🔐 Manual authentication for %s", p.targetDomain)
    p.logger.Println("Choose your method:")
    p.logger.Println("1. CloudFlared CLI (recommended)")
    p.logger.Println("2. Browser cookie extraction")

    reader := bufio.NewReader(os.Stdin)
    fmt.Print("\nChoose method (1 or 2): ")
    input, _ := reader.ReadString('\n')
    choice := strings.TrimSpace(input)

    if choice == "1" {
        p.logger.Println("\n📋 CloudFlared steps:")
        p.logger.Printf("1. Run: cloudflared access login https://%s", p.targetDomain)
        p.logger.Printf("2. Run: cloudflared access token -app=https://%s", p.targetDomain)
        p.logger.Println("3. Copy the token output and paste it below:")

        fmt.Print("\nCF_Authorization token: ")
        tokenInput, _ := reader.ReadString('\n')
        tokenValue := strings.TrimSpace(tokenInput)

        if tokenValue != "" {
            p.cfCookie = tokenValue
            // Set expiration to 23 hours from now (typical CF Access session)
            p.cookieExpires = time.Now().Unix() + int64(23*3600)
            return true
        }
    } else {
        p.logger.Println("\n📋 Browser cookie steps:")
        p.logger.Printf("1. Open your browser and go to: https://%s", p.targetDomain)
        p.logger.Println("2. Complete Cloudflare authentication")
        p.logger.Println("3. Open browser dev tools → Application → Cookies")
        p.logger.Printf("4. Find 'CF_Authorization' cookie for %s", p.targetDomain)
        p.logger.Println("5. Copy the cookie value and paste it here:")

        fmt.Print("\nCF_Authorization cookie: ")
        cookieInput, _ := reader.ReadString('\n')
        cookieValue := strings.TrimSpace(cookieInput)

        if cookieValue != "" {
            p.cfCookie = cookieValue
            // Set expiration to 23 hours from now (typical CF Access session)
            p.cookieExpires = time.Now().Unix() + int64(23*3600)
            return true
        }
    }

    return false
}

// isAuthenticated checks if we have a valid authentication cookie
func (p *CloudflareAccessProxy) isAuthenticated() bool {
    return p.cfCookie != "" && time.Now().Unix() < p.cookieExpires
}

// ensureAuthenticated ensures we have valid authentication
func (p *CloudflareAccessProxy) ensureAuthenticated() bool {
    if p.isAuthenticated() {
        return true
    }

    // Try cloudflared authentication first
    if p.authenticateWithCloudflared() {
        return true
    }

    // Try to get cookie from browser
    if p.extractCFCookieFromBrowser() {
        return true
    }

    // Fall back to manual authentication
    return p.authenticateManual()
}

// proxyHandler handles HTTP requests
type proxyHandler struct {
    proxy *CloudflareAccessProxy
}

// NewProxyHandler creates a new proxy handler
func NewProxyHandler(proxy *CloudflareAccessProxy) *proxyHandler {
    return &proxyHandler{proxy: proxy}
}

// ServeHTTP implements http.Handler interface
func (h *proxyHandler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
    // Ensure we're authenticated
    if !h.proxy.ensureAuthenticated() {
        http.Error(w, "Authentication failed", http.StatusUnauthorized)
        return
    }

    // Build target URL
    targetURL := fmt.Sprintf("https://%s%s", h.proxy.targetDomain, r.URL.RequestURI())

    // Create a new request
    req, err := http.NewRequest(r.Method, targetURL, nil)
    if err != nil {
        http.Error(w, "Failed to create request", http.StatusInternalServerError)
        return
    }

    // Copy headers from original request
    for name, values := range r.Header {
        for _, value := range values {
            req.Header.Add(name, value)
        }
    }

    // Remove host header as we're changing the target
    req.Header.Del("Host")

    // Remove accept-encoding to avoid compression issues
    req.Header.Del("Accept-Encoding")

    // Fix OCI manifest support for registry requests
    if strings.HasPrefix(r.URL.Path, "/v2/") {
        if accept := req.Header.Get("Accept"); accept != "" {
            ociTypes := []string{
                "application/vnd.oci.image.manifest.v1+json",
                "application/vnd.oci.image.index.v1+json",
            }
            for _, ociType := range ociTypes {
                if !strings.Contains(accept, ociType) {
                    req.Header.Set("Accept", accept+", "+ociType)
                }
            }
        }
    }

    // Add Cloudflare Access cookie
    cookieValue := fmt.Sprintf("CF_Authorization=%s", h.proxy.cfCookie)
    if existingCookie := req.Header.Get("Cookie"); existingCookie != "" {
        cookieValue = fmt.Sprintf("%s; %s", existingCookie, cookieValue)
    }
    req.Header.Set("Cookie", cookieValue)

    // Copy request body if present
    var body []byte
    if r.Body != nil {
        body, err = io.ReadAll(r.Body)
        if err != nil {
            http.Error(w, "Failed to read request body", http.StatusInternalServerError)
            return
        }
        req.Body = io.NopCloser(bytes.NewBuffer(body))
    }

    // Make the proxied request
    resp, err := h.proxy.session.Do(req)
    if err != nil {
        h.proxy.logger.Errorf("❌ Proxy error: %v", err)
        http.Error(w, fmt.Sprintf("Proxy error: %v", err), http.StatusInternalServerError)
        return
    }
    defer resp.Body.Close()

    // Copy response status
    w.WriteHeader(resp.StatusCode)

    // Copy response headers
    for name, values := range resp.Header {
        for _, value := range values {
            w.Header().Add(name, value)
        }
    }

    // Remove headers that could cause issues
    w.Header().Del("Transfer-Encoding")
    w.Header().Del("Connection")

    // Copy response body
    if r.Method != "HEAD" {
        _, err = io.Copy(w, resp.Body)
        if err != nil {
            h.proxy.logger.Errorf("❌ Failed to copy response body: %v", err)
        }
    }
}

// main function
func main() {
    // Parse command line arguments
    if len(os.Args) < 2 {
        fmt.Println("Usage: go run main.go <target_domain> [--port <port>]")
        os.Exit(1)
    }

    domain := os.Args[1]
    port := 8081

    if len(os.Args) > 3 && os.Args[2] == "--port" {
        p, err := strconv.Atoi(os.Args[3])
        if err != nil {
            fmt.Println("Invalid port number")
            os.Exit(1)
        }
        port = p
    }

    fmt.Printf("🚀 Starting Cloudflare Access Proxy\n")
    fmt.Printf("📡 Target: %s\n", domain)
    fmt.Printf("🔗 Local proxy: http://localhost:%d\n", port)

    // Create proxy instance
    proxy := NewCloudflareAccessProxy(domain, port)

    // Authenticate upfront before starting server
    fmt.Println("\n🔐 Authentication required before starting proxy...")
    if !proxy.ensureAuthenticated() {
        fmt.Println("❌ Authentication failed. Exiting.")
        os.Exit(1)
    }
    fmt.Println("✅ Authentication successful!")

    // Create HTTP server
    handler := NewProxyHandler(proxy)
    server := &http.Server{
        Addr:    fmt.Sprintf("localhost:%d", port),
        Handler: handler,
    }

    fmt.Printf("\n🚀 Proxy running on http://localhost:%d\n", port)
    fmt.Printf("💡 Configure your client to use localhost:%d instead of %s\n", port, domain)
    fmt.Println("🛑 Press Ctrl+C to stop")

    // Start server
    if err := server.ListenAndServe(); err != nil {
        fmt.Printf("❌ Server error: %v\n", err)
        os.Exit(1)
    }
}

// Helper function to check if a command exists
func commandExists(cmd string) bool {
    _, err := exec.LookPath(cmd)
    return err == nil
}
